src/fe-common/core/fe-common-core.c

```diff
@@ -462,25 +462,73 @@ void fe_common_core_finish_init(void)
 gboolean strarray_find_dest(char **array, const TEXT_DEST_REC *dest)
 {
 	g_return_val_if_fail(array != NULL, FALSE);
+	WI_ITEM_REC *item = window_item_find_window(dest->window, dest->server, dest->target);
+	int channel_type = module_get_uniq_id_str("WINDOW ITEM TYPE", "CHANNEL");
+	int query_type = module_get_uniq_id_str("WINDOW ITEM TYPE", "QUERY");
+	
+	char **tmp;
+	for (tmp = array; *tmp != NULL; tmp++) {
+		char *tagtarget = NULL;
+		if (g_ascii_strcasecmp(*tmp, "*") == 0)
+			return TRUE;
+		if (g_ascii_strcasecmp(*tmp, dest->target) == 0)
+			return TRUE;
 
-	if (strarray_find(array, "*") != -1)
-		return TRUE;
+		if (dest->server_tag != NULL) {
+			tagtarget = g_strdup_printf("%s/%s", dest->server_tag, "*");
+			if (tagtarget != NULL && g_ascii_strcasecmp(*tmp, tagtarget) == 0) {
+				g_free(tagtarget);
+				return TRUE;
+			}
+			g_free(tagtarget);
 
-	if (strarray_find(array, dest->target) != -1)
-		return TRUE;
+			tagtarget = g_strdup_printf("%s/%s", dest->server_tag, dest->target);
+			if (tagtarget != NULL && g_ascii_strcasecmp(*tmp, tagtarget) == 0) {
+				g_free(tagtarget);
+				return TRUE;
+			}
+			g_free(tagtarget);
+		}
 
-	if (dest->server_tag != NULL) {
-		char *tagtarget = g_strdup_printf("%s/%s", dest->server_tag, "*");
-		int ret = strarray_find(array, tagtarget);
-		g_free(tagtarget);
-		if (ret != -1)
-			return TRUE;
+		if (item->type && item->type == channel_type) {
+			if (g_ascii_strcasecmp(*tmp, "#") == 0)
+				return TRUE;
 
-		tagtarget = g_strdup_printf("%s/%s", dest->server_tag, dest->target);
-		ret = strarray_find(array, tagtarget);
-		g_free(tagtarget);
-		if (ret != -1)
-			return TRUE;
+			if (dest->server_tag != NULL)
+				tagtarget = g_strdup_printf("%s/%s", dest->server_tag, "#");
+			if (tagtarget != NULL && g_ascii_strcasecmp(*tmp, tagtarget) == 0) {
+				g_free(tagtarget);
+				return TRUE;
+			}
+			g_free(tagtarget);
+		}
+		else if (item->type && item->type == query_type) {
+			if (g_str_has_prefix(dest->target, "=")) {
```

Comment:
If you stare at the two branches of the `if` you can see that the code is the same, minus the "=" vs "@".
Can you fold the two branches together?